### PR TITLE
Pattern matching support for conditions and switch statements [NO SQUASH]

### LINKED ISSUE #157: Add support for pattern matching in conditions in CFG
Code like this should be correctly analyzed

```
public void Patterns(object o)
{
    if (o is null)
    {
        return;
    }

    if (o is int i)
    {
        Console.WriteLine(i);
    }
}
```

Apart from if statements, the other language elements that should support pattern matching are:
ternary operators, for loops, while loops, do loops

Switch statements are handled in the referenced ticket #1775

### LINKED ISSUE #1775: Add support for switch statements pattern matching in CFG
### Description

This is split from #157 and is related only to pattern matching in `switch` statements:

```
switch (o)
{
    case int n1:
        Console.Write(n1);
        break;
    case string s when s.Length > 0:
        Console.WriteLine(s);
        break;
    case null:
        Console.WriteLine("null");
        break;
}
```

### PR DESCRIPTION
**Pull request by valhristov**
Fix #157 
Fix #1775

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/CSharpControlFlowGraphBuilder.cs [hunk 1]
```diff
@@ -612,7 +642,12 @@ private Block BuildGotoStatement(GotoStatementSyntax statement, Block currentBlo
 
         #region Build switch
 
-        private Block BuildSwitchStatement(SwitchStatementSyntax switchStatement, Block currentBlock)
+        private Block BuildSwitchStatement(SwitchStatementSyntax switchStatement, Block currentBlock) =>
+            switchStatement.Sections.All(s => s.Labels.All(l => l.IsAnyKind(csharp6SwitchLabelKinds)))
+                ? BuildSwitchStatement_Legacy(switchStatement, currentBlock)
+                : BuildSwitchStatement_CSharp7(switchStatement, currentBlock);
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/CSharpControlFlowGraphBuilder.cs]: The pattern-enabled switch statements now takes into account the order of the case section declarations (the non-pattern switch doesn't), hence two completely different ways to handle.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/CSharpControlFlowGraphBuilder.cs [hunk 2]
```diff
@@ -442,6 +448,30 @@ private Block BuildExpression(ExpressionSyntax expression, Block currentBlock)
                             parent.ArgumentList?.Arguments.Select(a => a.Expression));
                     }
 
+                case SyntaxKindEx.IsPatternExpression:
+                    var isPatternExpression = (IsPatternExpressionSyntaxWrapper)expression;
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/CSharpControlFlowGraphBuilder.cs]: Would be nice to have some examples of the different cases.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ConditionEvaluatesToConstant.CSharp7.cs [hunk 3]
```diff
@@ -2,19 +2,103 @@
...
+                    s.ToString();
+                }
+            }
+        }
 
-        public bool Prop => TRUE || myBool; // Noncompliant
-                                            // Secondary@-1
+        void VariableDesignationPattern_Source(object o)
+        {
+            // We can set NotNull constraint only for one of the variables in the if condition
```
[the remark this answers is not in the document]
Comment (valhristov) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ConditionEvaluatesToConstant.CSharp7.cs]: To specify a relationship between two variables (e.g. `a > b` or `a != null`) we have a special SymbolicValue. To specify relationship between three variables (e.g. `a != null and b != null) we would need a new SV. Creating a new SV is difficult and not absolutely required to enable the new syntax. Perhaps there is another way, but I could not find it so far and I am open for ideas.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/SymbolicExecution/CSharpExplodedGraph.cs [hunk 4]
```diff
@@ -450,6 +451,33 @@ protected override void VisitInstruction(ExplodedGraphNode node)
...
+                    {
+                        // "x is string s"
+                        // VisitDeclarationPattern() expects SV_s on top of the stack, hence we pop SV_x
+                        newProgramState = newProgramState.PopValue();
+
+                        newProgramState = VisitDeclarationPattern((DeclarationPatternSyntaxWrapper)isPatternExpression.Pattern, newProgramState);
+                    }
+                    else
+                    {
+                        goto default;
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/SymbolicExecution/CSharpExplodedGraph.cs]: For a one line throw you can duplicate the code.

**Pull request merged**